From ed0cf0ef4ad43c3b9fa5374334e3bd9b1147ea7f Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Sat, 10 Sep 2016 14:45:14 +0000
Subject: [PATCH] OS-5662 vnd corrupts incoming MAC address frames Reviewed by:
 Cody Mello <melloc@joyent.com> Reviewed by: Patrick Mooney
 <patrick.mooney@joyent.com> Approved by: Patrick Mooney
 <patrick.mooney@joyent.com>

---
 usr/src/uts/common/io/vnd/vnd.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/usr/src/uts/common/io/vnd/vnd.c b/usr/src/uts/common/io/vnd/vnd.c
index 2abb6f9464..264212cfb4 100644
--- a/usr/src/uts/common/io/vnd/vnd.c
+++ b/usr/src/uts/common/io/vnd/vnd.c
@@ -2162,7 +2162,10 @@ vnd_mac_input(vnd_str_t *vsp, mac_resource_t *unused, mblk_t *mp_chain,
 		mp->b_rptr -= mhip->mhi_hdrsize;
 		vid = VLAN_ID(mhip->mhi_tci);
 		if (mhip->mhi_istagged && vid != VLAN_ID_NONE) {
-			bcopy(mp->b_rptr, mp->b_rptr + 4, 12);
+			/*
+			 * This is an overlapping copy. Do not use bcopy(9F).
+			 */
+			memmove(mp->b_rptr + 4, mp->b_rptr, 12);
 			mp->b_rptr += 4;
 		}
 
@@ -2180,7 +2183,6 @@ vnd_mac_input(vnd_str_t *vsp, mac_resource_t *unused, mblk_t *mp_chain,
 		signal |= vnd_dq_push(&vsp->vns_dq_read, mp, B_FALSE,
 		    vnd_drop_in);
 		mutex_exit(&vsp->vns_dq_read.vdq_lock);
-
 	}
 
 	if (signal != 0) {
-- 
2.21.0

